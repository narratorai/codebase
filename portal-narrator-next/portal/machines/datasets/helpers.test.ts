import _ from 'lodash'
import {
  getIftttColumnType,
  handleUpdateComputationColumn,
  makeColumnOptionsWithExistingLabels,
  makeMetricsOnPivotedToggle,
  reverseMetricComputeColumns,
} from './helpers'

import datasetContext from '../../../test/fixtures/dataset/datasetContext.json'
import datasetWithGroupIftttContext from '../../../test/fixtures/dataset/datasetWithGroupIfttt.json'
import datasetGroupResponse from '../../../test/fixtures/datasetGroupResponse.json'

import addComputColumnEventParent from '../../../test/fixtures/datasetMachine/addComputeColumnEventParent.json'
import groupFuncValidationInParentContext from '../../../test/fixtures/datasetMachine/groupFuncValidationInParentContext.json'

import addComputColumnEventWithGroupFuncInGroup from '../../../test/fixtures/datasetMachine/addComputeColumnEventWithGroupFuncInGroup.json'
import groupFuncValidatedInGroupContext from '../../../test/fixtures/datasetMachine/groupFuncValidatedInGroupContext.json'

import {
  DatasetContext,
  DatasetEvent,
  IDatasetQueryColumn,
  IDatasetQueryGroup,
  IDatasetQueryGroupComputedColumn,
} from 'util/datasets/interfaces'

describe('#makeColumnOptionsWithExistingLabels', () => {
  describe('for cohort activity options', () => {
    it('overwrites labels from api with labels from formValue', () => {
      const options = makeColumnOptionsWithExistingLabels({
        options: [
          {
            dropdown_label: '1',
            name: 'ts',
            label: 'Ts',
            type: 'timestamp',
            enrichment_table: null,
            values: [{ key: 'independent_completed_order', value: '43.88%' }],
          },
          {
            dropdown_label: '2',
            name: 'feature_2',
            label: 'Page Path',
            type: 'string',
            enrichment_table: null,
            values: [],
          },
        ],
        formValue: {
          cohort: {
            columns: [{ name: 'ts', label: 'started_session_at', type: 'timestamp', enrichment_table: null }],
          },
        },
      })
      expect(options).toMatchSnapshot()
    })
  })

  describe('for append activity/join options', () => {
    it('overwrites labels from api with labels from formValue', () => {
      const options = makeColumnOptionsWithExistingLabels({
        options: [
          { dropdown_label: '1', name: 'ts', label: 'Ts', type: 'timestamp', enrichment_table: null, values: [] },
          {
            dropdown_label: '2',
            name: 'feature_2',
            label: 'Page Path',
            type: 'string',
            enrichment_table: null,
            values: [],
          },
        ],
        appendActivityIndex: 0,
        formValue: {
          append_activities: [
            {
              columns: [{ name: 'ts', label: 'started_session_at', type: 'timestamp', enrichment_table: null }],
            },
          ],
        },
      })
      expect(options).toMatchSnapshot()
    })
  })
})

const pivotValues = ['first_value', 'second_value,', 'third_value']

describe('#makeMetricsOnPivotedToggle', () => {
  it('creates a metric column for each distinct value', () => {
    const group = datasetGroupResponse.groupings.group_by_kind_and_source.query_definition.query
      .all_groups[0] as unknown as IDatasetQueryGroup

    expect(
      makeMetricsOnPivotedToggle({
        columnId: group.columns[0].id,
        group,
        pivotValues,
      })
    ).toMatchSnapshot()
  })
})

describe('#reverseMetricComputeColumns', () => {
  it('reverses the pivot metrics', () => {
    const groupQuery = datasetGroupResponse.groupings.group_by_kind_and_source.query_definition.query
      .all_groups[0] as unknown as IDatasetQueryGroup

    const { pivotedMetrics, pivotedComputedColumns } = makeMetricsOnPivotedToggle({
      columnId: groupQuery.columns[0].id,
      group: groupQuery,
      pivotValues,
    })

    const updatedGroup = {
      ...groupQuery,
      columns: [{ ...groupQuery.columns[0], pivoted: true }, groupQuery.columns[1]],
      computed_columns: [...groupQuery.computed_columns, ...pivotedComputedColumns],
      metrics: [...groupQuery.metrics, ...pivotedMetrics],
      order: [],
    }

    expect(reverseMetricComputeColumns({ group: updatedGroup })).toMatchSnapshot()
  })
})

describe('#getIftttColumnType', () => {
  // @ts-ignore: datasetContext is directly copied from an example dataset - ignoring for now
  const context = datasetContext as DatasetContext

  it("returns column's 'value_kind' if not 'column_id' or 'null'", () => {
    const stringElseValueColumn = _.find(datasetContext.columns, ['id', 'else_string']) as
      | IDatasetQueryGroupComputedColumn
      | IDatasetQueryColumn
    const type = getIftttColumnType({ column: stringElseValueColumn, context })

    expect(type).toBe('string')
  })

  it("returns column's referenced column_id's value_kind if column's value_kind is 'column_id'", () => {
    const columnWithColumnId = _.find(datasetContext.columns, ['id', 'else_column_id_timestamp']) as
      | IDatasetQueryGroupComputedColumn
      | IDatasetQueryColumn
    const type = getIftttColumnType({ column: columnWithColumnId, context })

    expect(type).toBe('timestamp')
  })

  it("returns first non null case's value_kind if column's value_kind is null", () => {
    const columnWithCase = _.find(datasetContext.columns, ['id', 'broken_null_column__qeXeYQ3y']) as
      | IDatasetQueryGroupComputedColumn
      | IDatasetQueryColumn
    const type = getIftttColumnType({ column: columnWithCase, context })

    expect(type).toBe('number')
  })

  it("returns first non null case's referenced column_ids value_kind if column's value_kind is null and case's value_kind is column_id", () => {
    const columnWithCaseWithColumnId = _.find(datasetContext.columns, ['id', 'broken_null_column']) as
      | IDatasetQueryGroupComputedColumn
      | IDatasetQueryColumn
    const type = getIftttColumnType({ column: columnWithCaseWithColumnId, context })

    expect(type).toBe('string')
  })

  // @ts-ignore: datasetContext is directly copied from an example dataset - ignoring for now
  const withGroupContext = datasetWithGroupIftttContext as DatasetContext

  it('returns a type for a group column', () => {
    // The first group in withGroupContext
    const groupSlug = _.get(withGroupContext, 'all_groups[0].slug')
    // has an ifttt as the first computed column
    const column = _.get(withGroupContext, 'all_groups[0].computed_columns[0]')

    const type = getIftttColumnType({ column: column, context: withGroupContext, groupSlug })
    expect(type).toBe('number')
  })
})

describe('#handleUpdateComputationColumn', () => {
  describe('in a parent tab', () => {
    it('creates a compute column with group_func', () => {
      const computationColumns = handleUpdateComputationColumn({
        context: groupFuncValidationInParentContext as unknown as DatasetContext,
        event: addComputColumnEventParent as DatasetEvent,
      })

      // Check each compute column
      // and ignore "id" since it is always unique
      // https://stackoverflow.com/questions/51619100/using-jest-property-matchers-on-arrays-of-objects
      computationColumns?.forEach((computeColumn) => {
        expect(computeColumn).toMatchSnapshot({
          id: expect.any(String),
        })
      })
    })
  })

  describe('in a group tab', () => {
    it('creates a compute column with group_func', () => {
      const group = handleUpdateComputationColumn({
        context: groupFuncValidatedInGroupContext as unknown as DatasetContext,
        event: addComputColumnEventWithGroupFuncInGroup as DatasetEvent,
      })?.[0] as IDatasetQueryGroup

      // Test group without compute seperately b/c computed columns have unique ids
      const groupWithoutComputeColumns = _.omit(group, 'computed_columns')
      expect(groupWithoutComputeColumns).toMatchSnapshot()

      // check each compute column
      // and ignore "id" since it is always unique
      // https://stackoverflow.com/questions/51619100/using-jest-property-matchers-on-arrays-of-objects
      const computationColumns = group?.computed_columns
      computationColumns?.forEach((computeColumn) => {
        expect(computeColumn).toMatchSnapshot({
          id: expect.any(String),
        })
      })
    })
  })
})
