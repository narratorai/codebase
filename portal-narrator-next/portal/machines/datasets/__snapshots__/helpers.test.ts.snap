// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#handleUpdateComputationColumn in a group tab creates a compute column with group_func 1`] = `
{
  "columns": [
    {
      "column_id": "feature_2c7f9cff7",
      "column_kind": "group",
      "filters": [],
      "id": "feature_1_2ecde358",
      "label": "Feature 1",
      "mavis_type": "string",
      "name": "",
      "output": true,
      "pivoted": false,
      "source_kind": "group",
      "type": "string",
    },
    {
      "column_id": "feature_3b508423f",
      "column_kind": "group",
      "filters": [],
      "id": "feature_2_5b9c9450",
      "label": "Feature 2",
      "mavis_type": "string",
      "name": "",
      "output": true,
      "pivoted": false,
      "source_kind": "group",
      "type": "string",
    },
    {
      "column_id": "link7e16eb22",
      "column_kind": "group",
      "filters": [],
      "id": "link_14a609c0",
      "label": "Link",
      "mavis_type": "string",
      "name": "",
      "output": true,
      "pivoted": false,
      "source_kind": "group",
      "type": "string",
    },
  ],
  "metrics": [
    {
      "agg_function": "COUNT_ALL",
      "column_id": null,
      "column_kind": "metrics",
      "conditioned_on_columns": null,
      "filters": [],
      "id": "total_marketing_session_rows_17fc7f48",
      "label": "Total Marketing Session Rows",
      "mavis_type": null,
      "output": true,
      "pivot": [],
      "type": "integer",
    },
    {
      "agg_function": "SUM",
      "column_id": "did_email_clickedddf9679c",
      "column_kind": "metrics",
      "conditioned_on_columns": null,
      "filters": [],
      "id": "total_email_clicked_between_49d8fc30",
      "label": "Total Email Clicked Between",
      "mavis_type": null,
      "output": true,
      "pivot": [],
      "type": "float",
    },
    {
      "agg_function": "AVERAGE",
      "column_id": "did_email_clickedddf9679c",
      "column_kind": "metrics",
      "conditioned_on_columns": null,
      "filters": [],
      "id": "conversion_rate_to_email_clicked_between_585ba8ac",
      "label": "Conversion Rate to Email Clicked Between",
      "mavis_type": null,
      "output": true,
      "pivot": [],
      "type": "float",
    },
    {
      "agg_function": "AVERAGE",
      "column_id": "day_email_clicked364ea5b7",
      "column_kind": "metrics",
      "conditioned_on_columns": null,
      "filters": [],
      "id": "average_days_to_email_clicked_6e07ad0f",
      "label": "Average Days To Email Clicked",
      "mavis_type": null,
      "output": true,
      "pivot": [],
      "type": "float",
    },
  ],
  "name": "by Feature 1 & Feature 2 & Link",
  "order": [
    {
      "column_id": "total_marketing_session_rows_17fc7f48",
      "order_direction": "asc",
    },
  ],
  "parent_filters": [],
  "pivot": [],
  "slug": "feature_1_feature_2_link11a8f00b",
}
`;

exports[`#handleUpdateComputationColumn in a group tab creates a compute column with group_func 2`] = `
{
  "_auto_generated_by": null,
  "filters": [],
  "group_func": [],
  "id": Any<String>,
  "label": "Group Func in Group",
  "name": "",
  "output": true,
  "source_details": {
    "kind": "freehand_function",
    "raw_string": "iff(not_is_null(link_14a609c0), '1', '0')",
  },
  "source_kind": "computed",
  "type": "string",
}
`;

exports[`#handleUpdateComputationColumn in a parent tab creates a compute column with group_func 1`] = `
{
  "filters": [],
  "group_func": null,
  "id": Any<String>,
  "label": "Activity Id",
  "mavis_type": "string",
  "name": "activity_id",
  "output": true,
  "source_details": {
    "activity_id": "marketing_sessionbd7670fb",
    "activity_kind": "limiting",
    "applied_function": null,
    "enrichment_table": null,
  },
  "source_kind": "activity",
  "type": "string",
}
`;

exports[`#handleUpdateComputationColumn in a parent tab creates a compute column with group_func 2`] = `
{
  "filters": [],
  "group_func": null,
  "id": Any<String>,
  "label": "Timestamp",
  "mavis_type": "timestamp",
  "name": "ts",
  "output": true,
  "source_details": {
    "activity_id": "marketing_sessionbd7670fb",
    "activity_kind": "limiting",
    "applied_function": null,
    "enrichment_table": null,
  },
  "source_kind": "activity",
  "type": "timestamp",
}
`;

exports[`#handleUpdateComputationColumn in a parent tab creates a compute column with group_func 3`] = `
{
  "filters": [],
  "group_func": null,
  "id": Any<String>,
  "label": "Customer",
  "mavis_type": "string",
  "name": "customer",
  "output": true,
  "source_details": {
    "activity_id": "marketing_sessionbd7670fb",
    "activity_kind": "limiting",
    "applied_function": null,
    "enrichment_table": null,
  },
  "source_kind": "activity",
  "type": "string",
}
`;

exports[`#handleUpdateComputationColumn in a parent tab creates a compute column with group_func 4`] = `
{
  "filters": [],
  "group_func": null,
  "id": Any<String>,
  "label": "Unique Identifier",
  "mavis_type": "string",
  "name": "join_customer",
  "output": true,
  "source_details": {
    "activity_id": "marketing_sessionbd7670fb",
    "activity_kind": "limiting",
    "applied_function": null,
    "enrichment_table": null,
  },
  "source_kind": "activity",
  "type": "string",
}
`;

exports[`#handleUpdateComputationColumn in a parent tab creates a compute column with group_func 5`] = `
{
  "filters": [],
  "group_func": null,
  "id": Any<String>,
  "label": "Activity",
  "mavis_type": "string",
  "name": "feature_1",
  "output": true,
  "source_details": {
    "activity_id": "marketing_sessionbd7670fb",
    "activity_kind": "limiting",
    "applied_function": null,
    "enrichment_table": null,
  },
  "source_kind": "activity",
  "type": "string",
}
`;

exports[`#handleUpdateComputationColumn in a parent tab creates a compute column with group_func 6`] = `
{
  "filters": [],
  "group_func": null,
  "id": Any<String>,
  "label": "Feature 1",
  "mavis_type": "string",
  "name": "feature_2",
  "output": true,
  "source_details": {
    "activity_id": "marketing_sessionbd7670fb",
    "activity_kind": "limiting",
    "applied_function": null,
    "enrichment_table": null,
  },
  "source_kind": "activity",
  "type": "string",
}
`;

exports[`#handleUpdateComputationColumn in a parent tab creates a compute column with group_func 7`] = `
{
  "filters": [],
  "group_func": null,
  "id": Any<String>,
  "label": "Feature 2",
  "mavis_type": "string",
  "name": "feature_3",
  "output": true,
  "source_details": {
    "activity_id": "marketing_sessionbd7670fb",
    "activity_kind": "limiting",
    "applied_function": null,
    "enrichment_table": null,
  },
  "source_kind": "activity",
  "type": "string",
}
`;

exports[`#handleUpdateComputationColumn in a parent tab creates a compute column with group_func 8`] = `
{
  "filters": [],
  "group_func": null,
  "id": Any<String>,
  "label": "Link",
  "mavis_type": "string",
  "name": "link",
  "output": true,
  "source_details": {
    "activity_id": "marketing_sessionbd7670fb",
    "activity_kind": "limiting",
    "applied_function": null,
    "enrichment_table": null,
  },
  "source_kind": "activity",
  "type": "string",
}
`;

exports[`#handleUpdateComputationColumn in a parent tab creates a compute column with group_func 9`] = `
{
  "filters": [],
  "group_func": null,
  "id": Any<String>,
  "label": "Activity Occurrence",
  "mavis_type": "integer",
  "name": "activity_occurrence",
  "output": true,
  "source_details": {
    "activity_id": "marketing_sessionbd7670fb",
    "activity_kind": "limiting",
    "applied_function": null,
    "enrichment_table": null,
  },
  "source_kind": "activity",
  "type": "number",
}
`;

exports[`#handleUpdateComputationColumn in a parent tab creates a compute column with group_func 10`] = `
{
  "filters": [],
  "group_func": null,
  "id": Any<String>,
  "label": "First In Between Email Clicked Timestamp",
  "mavis_type": "timestamp",
  "name": "ts",
  "output": true,
  "source_details": {
    "activity_id": "email_clicked6a87a66d",
    "activity_kind": "append",
    "applied_function": null,
    "enrichment_table": null,
  },
  "source_kind": "activity",
  "type": "timestamp",
}
`;

exports[`#handleUpdateComputationColumn in a parent tab creates a compute column with group_func 11`] = `
{
  "filters": [],
  "group_func": [
    "sum",
    "average",
  ],
  "id": Any<String>,
  "label": "Did Email Clicked Between",
  "mavis_type": "integer",
  "name": null,
  "output": true,
  "source_details": {
    "activity_id": "email_clicked6a87a66d",
    "activity_kind": "append",
    "enrichment_table": null,
    "raw_string": "exists(tsabb33d19)",
  },
  "source_kind": "computed",
  "type": "integer",
}
`;

exports[`#handleUpdateComputationColumn in a parent tab creates a compute column with group_func 12`] = `
{
  "filters": [],
  "group_func": [
    "average",
  ],
  "id": Any<String>,
  "label": "Days To Email Clicked",
  "mavis_type": "timestamp",
  "name": null,
  "output": true,
  "source_details": {
    "activity_id": "email_clicked6a87a66d",
    "activity_kind": "append",
    "enrichment_table": null,
    "raw_string": "time_diff('day', join_ts.local, tsabb33d19)",
  },
  "source_kind": "computed",
  "type": "integer",
}
`;

exports[`#handleUpdateComputationColumn in a parent tab creates a compute column with group_func 13`] = `
{
  "_auto_generated_by": null,
  "filters": [],
  "group_func": [],
  "id": Any<String>,
  "label": "Group Func in Parent",
  "name": "",
  "output": true,
  "source_details": {
    "kind": "freehand_function",
    "raw_string": "iff(not_is_null(linka0da20a2), '1', '0')",
  },
  "source_kind": "computed",
  "type": "string",
}
`;

exports[`#makeColumnOptionsWithExistingLabels for append activity/join options overwrites labels from api with labels from formValue 1`] = `
[
  {
    "enrichment_table": null,
    "label": "started_session_at",
    "name": "ts",
    "type": "timestamp",
    "values": [],
  },
  {
    "dropdown_label": "2",
    "enrichment_table": null,
    "label": "Page Path",
    "name": "feature_2",
    "type": "string",
    "values": [],
  },
]
`;

exports[`#makeColumnOptionsWithExistingLabels for cohort activity options overwrites labels from api with labels from formValue 1`] = `
[
  {
    "enrichment_table": null,
    "label": "started_session_at",
    "name": "ts",
    "type": "timestamp",
    "values": [
      {
        "key": "independent_completed_order",
        "value": "43.88%",
      },
    ],
  },
  {
    "dropdown_label": "2",
    "enrichment_table": null,
    "label": "Page Path",
    "name": "feature_2",
    "type": "string",
    "values": [],
  },
]
`;

exports[`#makeMetricsOnPivotedToggle creates a metric column for each distinct value 1`] = `
{
  "pivotedComputedColumns": [
    {
      "_auto_generated_by": "create-pivot",
      "filters": [],
      "id": "metric_total_ordered_impression_kit_first_value__computed_conversion",
      "label": "Conversion Rate to total_ordered_impression_kit first_value",
      "name": "",
      "output": true,
      "source_details": {
        "kind": "freehand_function",
        "raw_string": "({metric_total_ordered_impression_kit_first_value}*1.0)/{metrics_total_events_first_value}",
      },
      "source_kind": "computed",
      "type": "float",
    },
    {
      "_auto_generated_by": "create-pivot",
      "filters": [],
      "id": "metric_total_ordered_impression_kit_second_value___computed_conversion",
      "label": "Conversion Rate to total_ordered_impression_kit second_value,",
      "name": "",
      "output": true,
      "source_details": {
        "kind": "freehand_function",
        "raw_string": "({metric_total_ordered_impression_kit_second_value_}*1.0)/{metrics_total_events_second_value_}",
      },
      "source_kind": "computed",
      "type": "float",
    },
    {
      "_auto_generated_by": "create-pivot",
      "filters": [],
      "id": "metric_total_ordered_impression_kit_third_value__computed_conversion",
      "label": "Conversion Rate to total_ordered_impression_kit third_value",
      "name": "",
      "output": true,
      "source_details": {
        "kind": "freehand_function",
        "raw_string": "({metric_total_ordered_impression_kit_third_value}*1.0)/{metrics_total_events_third_value}",
      },
      "source_kind": "computed",
      "type": "float",
    },
  ],
  "pivotedMetrics": [
    {
      "_auto_generated_by": "create-pivot",
      "_pre_pivot_column_id": "metrics_total_events",
      "_pre_pivot_column_label": "total_events",
      "agg_function": "COUNT",
      "column_id": "*",
      "filters": [],
      "id": "metrics_total_events_first_value",
      "label": "total_events first_value",
      "output": true,
      "pivot": [
        {
          "column_id": "group_session_session_kind",
          "value": "first_value",
        },
      ],
      "type": "number",
    },
    {
      "_auto_generated_by": "create-pivot",
      "_pre_pivot_column_id": "metrics_total_events",
      "_pre_pivot_column_label": "total_events",
      "agg_function": "COUNT",
      "column_id": "*",
      "filters": [],
      "id": "metrics_total_events_second_value_",
      "label": "total_events second_value,",
      "output": true,
      "pivot": [
        {
          "column_id": "group_session_session_kind",
          "value": "second_value,",
        },
      ],
      "type": "number",
    },
    {
      "_auto_generated_by": "create-pivot",
      "_pre_pivot_column_id": "metrics_total_events",
      "_pre_pivot_column_label": "total_events",
      "agg_function": "COUNT",
      "column_id": "*",
      "filters": [],
      "id": "metrics_total_events_third_value",
      "label": "total_events third_value",
      "output": true,
      "pivot": [
        {
          "column_id": "group_session_session_kind",
          "value": "third_value",
        },
      ],
      "type": "number",
    },
    {
      "_auto_generated_by": "create-pivot",
      "_pre_pivot_column_id": "metric_total_ordered_impression_kit",
      "_pre_pivot_column_label": "total_ordered_impression_kit",
      "agg_function": "SUM",
      "column_id": "converted_to_ordered_impression_kit",
      "filters": [],
      "id": "metric_total_ordered_impression_kit_first_value",
      "label": "total_ordered_impression_kit first_value",
      "output": true,
      "pivot": [
        {
          "column_id": "group_session_session_kind",
          "value": "first_value",
        },
      ],
      "type": "number",
    },
    {
      "_auto_generated_by": "create-pivot",
      "_pre_pivot_column_id": "metric_total_ordered_impression_kit",
      "_pre_pivot_column_label": "total_ordered_impression_kit",
      "agg_function": "SUM",
      "column_id": "converted_to_ordered_impression_kit",
      "filters": [],
      "id": "metric_total_ordered_impression_kit_second_value_",
      "label": "total_ordered_impression_kit second_value,",
      "output": true,
      "pivot": [
        {
          "column_id": "group_session_session_kind",
          "value": "second_value,",
        },
      ],
      "type": "number",
    },
    {
      "_auto_generated_by": "create-pivot",
      "_pre_pivot_column_id": "metric_total_ordered_impression_kit",
      "_pre_pivot_column_label": "total_ordered_impression_kit",
      "agg_function": "SUM",
      "column_id": "converted_to_ordered_impression_kit",
      "filters": [],
      "id": "metric_total_ordered_impression_kit_third_value",
      "label": "total_ordered_impression_kit third_value",
      "output": true,
      "pivot": [
        {
          "column_id": "group_session_session_kind",
          "value": "third_value",
        },
      ],
      "type": "number",
    },
  ],
}
`;

exports[`#reverseMetricComputeColumns reverses the pivot metrics 1`] = `
{
  "computedColumns": [],
  "metrics": [
    {
      "agg_function": "COUNT",
      "column_id": "*",
      "filters": [],
      "id": "metrics_total_events",
      "label": "total_events",
      "output": true,
      "pivot": [],
      "type": "number",
    },
    {
      "agg_function": "SUM",
      "column_id": "converted_to_ordered_impression_kit",
      "filters": [],
      "id": "metric_total_ordered_impression_kit",
      "label": "total_ordered_impression_kit",
      "output": true,
      "pivot": [],
      "type": "number",
    },
  ],
}
`;
